File: D1L1/DateFormat.hs
module DateFormat where
import Text.Printf
import Data.Maybe

data Date = Date { day :: Int
                 , month :: Int } 
instance Eq Date where
    (==) (Date d1 m1) (Date d2 m2) = d1 == d2 && m1 == m2
instance Show Date where
    show (Date d m) = show' d ++ "/" ++ show' m
        where show' = printf "%02d"
instance Ord Date where
    (<) (Date d1 m1) (Date d2 m2) = m1 < m2 || (m1 == m2 && d1 < d2)
    (<=) (Date d1 m1) (Date d2 m2) = m1 < m2 || (m1 == m2 && d1 <= d2)

toDate :: String -> Date
toDate [a,b,c,d,e] = Date (read [a,b]) (read [d,e])

rotateDate :: Date -> Date
rotateDate (Date d m) = Date m d

validDate :: Date -> Bool
validDate (Date d m) = m <= 12
    
fixDate :: Maybe Date -> Date -> Maybe Date
fixDate Nothing _ = Nothing
fixDate (Just prev) cur = 
    if prev < cur then if validDate temp && temp < cur && prev < temp 
                           then Just temp
                           else if validDate cur then Just cur
                                                 else Nothing
                  else if validDate temp && prev < temp 
                           then Just temp
                           else Nothing
    where temp = rotateDate cur
    
fixDates :: [Date] -> String
fixDates ds = if length xs == length ds 
                 then unwords . map (show . rotateDate . fromJust) $ xs
                 else ""
   where xs = tail . takeWhile isJust . scanl fixDate (Just (Date 0 0)) $ ds

fromEuropeanToUs :: [String] -> String
fromEuropeanToUs = fixDates . map toDate . words . concat

File: D1L1/DateFormat.cpp
#include <vector>
#include <string>
#include <sstream>
#include <iomanip>
using namespace std;

class Date {
public:
    int day, month; 
    Date (int day, int month) : day(day), month(month) { } 
    bool operator < (Date const &date) const {
        if (month < date.month) return true;
        if (month > date.month) return false;
        return day < date.day;
    }
    bool operator <= (Date const &date) const { 
        if (month < date.month) return true;
        if (month > date.month) return false;
        return day <= date.day;
    }
    Date& operator = (Date const &date) {
        day = date.day;
        month = date.month;
        return *this;
    }
    bool valid() const {
        return month <= 12;
    }
}; 
istream& operator >> (istream &is, Date &date) {
    is >> date.day; is.ignore(1); is >> date.month;
    return is;
}
ostream& operator << (ostream &os, Date const &date) {
    os << setw(2) << setfill('0') << date.month << '/' 
       << setw(2) << setfill('0') << date.day;
    return os;
}

string concat(vector<string> v) {
    ostringstream oss;
    for (int i = 0; i < v.size(); ++i)
        oss << v[i];
    return oss.str();
} 
string fromEuropeanToUs(vector<string> const &dateList)
{
    istringstream iss(concat(dateList)); ostringstream oss; 
    Date prev(0, 0), cur(0, 0);
    while (true) { 
        iss >> cur; 
        Date temp(cur.month, cur.day);

        if (cur <= prev) {
            if (temp <= cur || temp <= prev) return "";
            cur = temp;
        }
        else if (prev < temp && temp < cur) cur = temp; 
        if (!cur.valid()) return "";

        oss << cur; 
        if (iss.good()) { 
            oss << ' '; 
            prev = cur;
        }
        else break;
    } 
    return oss.str();
}
